src/compute/src/row_spine.rs

```diff
@@ -0,0 +1,310 @@
+        RcBuilder<OrdValBuilder<RowValLayout<((Row, V), T, R)>>>,
+    >;
+    pub type RowSpine<T, R> = Spine<
+        Rc<OrdKeyBatch<RowLayout<((Row, ()), T, R)>>>,
+        ColumnatedMergeBatcher<Row, (), T, R>,
+        RcBuilder<OrdKeyBuilder<RowLayout<((Row, ()), T, R)>>>,
+    >;
+
+    /// A layout based on timely stacks
+    pub struct RowRowLayout<U: Update<Key = Row, Val = Row>> {
+        phantom: std::marker::PhantomData<U>,
+    }
+    pub struct RowValLayout<U: Update<Key = Row>> {
+        phantom: std::marker::PhantomData<U>,
+    }
+    pub struct RowLayout<U: Update<Key = Row, Val = ()>> {
+        phantom: std::marker::PhantomData<U>,
+    }
+
+    impl<U: Update<Key = Row, Val = Row>> Layout for RowRowLayout<U>
+    where
+        U::Time: Columnation,
+        U::Diff: Columnation,
+    {
+        type Target = U;
+        type KeyContainer = DatumContainer;
+        type ValContainer = DatumContainer;
+        type UpdContainer = TimelyStack<(U::Time, U::Diff)>;
+    }
+    impl<U: Update<Key = Row>> Layout for RowValLayout<U>
+    where
+        U::Val: Columnation,
+        U::Time: Columnation,
+        U::Diff: Columnation,
+    {
+        type Target = U;
+        type KeyContainer = DatumContainer;
+        type ValContainer = TimelyStack<U::Val>;
+        type UpdContainer = TimelyStack<(U::Time, U::Diff)>;
+    }
+    impl<U: Update<Key = Row, Val = ()>> Layout for RowLayout<U>
+    where
+        U::Time: Columnation,
+        U::Diff: Columnation,
+    {
+        type Target = U;
+        type KeyContainer = DatumContainer;
+        type ValContainer = TimelyStack<()>;
+        type UpdContainer = TimelyStack<(U::Time, U::Diff)>;
+    }
+}
+
+/// A `Row`-specialized container using dictionary compression.
+mod container {
+
+    use differential_dataflow::trace::cursor::MyTrait;
+    use differential_dataflow::trace::implementations::BatchContainer;
+    use differential_dataflow::trace::implementations::OffsetList;
+
+    use mz_repr::{read_datum, Datum, Row};
+
+    /// A slice container with four bytes overhead per slice.
+    pub struct DatumContainer {
+        batches: Vec<DatumBatch>,
+    }
+
+    impl DatumContainer {
+        /// Visit contained allocations to determine their size and capacity.
+        #[inline]
+        pub fn heap_size(&self, mut callback: impl FnMut(usize, usize)) {
+            // Calculate heap size for local, stash, and stash entries
+            callback(
+                self.batches.len() * std::mem::size_of::<DatumBatch>(),
+                self.batches.capacity() * std::mem::size_of::<DatumBatch>(),
+            );
+            for batch in self.batches.iter() {
+                callback(
+                    batch.offsets.smol.len() * 4,
+                    batch.offsets.smol.capacity() * 4,
+                );
+                callback(
+                    batch.offsets.chonk.len() * 8,
+                    batch.offsets.chonk.capacity() * 8,
+                );
```

Comment:
You could use `offset_list_size` here: https://github.com/MaterializeInc/materialize/pull/23838/files#diff-b08a74c4ed149bcdb4bf9c987e6ded9bcf39931394b3113abf70347173dacd09R229

Reply:
Done!